### PR TITLE
JIT: Invalidate m_switchDescMap instead of updating it

### LINKED ISSUE #98772: Diff in checked vs. release
Call to `GetSwitchDescMap()` that creates `m_switchDescMap` if it is nullptr. `m_switchDescMap` is invalidated everytime we renumber blocks or we replace jump targets. However, if I trace the caller of `GetSwitchDescMap()`, I see that it can get called from DEBUG via `fgDebugCheckProfileWeights()` and repopulate the map, which will behave differently at later point of time when we `UpdateSwitchTableTarget()`.

https://github.com/dotnet/runtime/blob/703f6e846027681a052970929989bd36e81636cd/src/coreclr/jit/fgflow.cpp#L627-L630

In Debug, `map != nullptr` and hence we update the target, but in Release, it is still nullptr. In Debug, since we updated the target, it changes the successors we visit and variables created, offsets they get assigned to, etc.

image

Mostly caused by https://github.com/dotnet/runtime/pull/98526

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Fixes #98772. Recently, we started checking successor edge likelihoods in `Compiler::fgDebugCheckProfileWeights` by default; this means we call `Compiler::fgDebugCheckOutgoingProfileData` in Debug/Checked builds to verify successor edges' likelihoods, and thus call `BasicBlock::GetSucc(unsigned, Compiler*)` to iterate the successor edges. For switch blocks, `GetSucc(unsigned, Compiler*)` calls `GetSwitchDescMap`, and builds `m_switchDescMap` if it doesn't exist yet. Upon finishing edge likelihood verification, we don't reset `m_switchDescMap`, so it is possible for this map to be created earlier in Debug/Checked builds versus Release builds.

This doesn't result in behavioral diffs if the map contains the same state between Debug/Release builds by the time it is read. However, if the map is null by the time it is needed, it is created on-demand, ensuring it is completely up-to-date. If the map is not null, the correctness of its current state depends on how judiciously it was maintained with `UpdateSwitchTableTarget`, creating potential for diffs in the map's state. By invalidating the map instead of updating it as state changes, we can force it to be rebuilt with the latest state when it's needed.

### CODE REVIEW

File: src/coreclr/jit/fgopt.cpp [hunk 1]
```diff
@@ -1871,20 +1871,23 @@ bool Compiler::fgOptimizeSwitchBranches(BasicBlock* block)
...
+            modified = true;
 
             goto REPEAT_SWITCH;
         }
     } while (++jmpTab, --jmpCnt);
 
+    if (modified)
+    {
+        // Invalidate the set of unique targets for block, since we modified the targets
+        InvalidateUniqueSwitchSuccMap();
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgopt.cpp:1888]:
```suggestion
        fgInvalidateUniqueSwitchSuccMapEntry(block);
```
? Or do we need to invalidate the full thing?

Comment (amanasifkhalid) [reply src/coreclr/jit/fgopt.cpp]: That seems to work; fixed.

**Review by jakobbotsch — APPROVED** [review]

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib for visibility.

**Comment by amanasifkhalid** [comment]
`runtime-coreclr superpmi-asmdiffs-checked-release` [passed](https://dev.azure.com/dnceng-public/public/_build/results?buildId=575881&view=results).

**Pull request merged**